rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+   destroyed_at_s: u64;
+
+
+
+
+   ...
+   // A list of current checkpoints that may be active
+   checkpoint: [Checkpoints] (required);
+}
+
+
+table WriterCheckpoint {
+   epoch: u64;
+}
+
+union CheckpointMetadata { WriterCheckpointMetadata }
+
+
+table Uuid {
+   high: uint64;
+   low: uint64;
+}
+
+
+// Checkpoint reference to be included in manifest to record active checkpoints.
+table Checkpoint {
+   // Id that must be unique across all open checkpoints.
+   id: Uuid (required);
+
+
+   // The manifest ID that this checkpoint is using as its `CoreDbState`.
+   manifest_id: ulong;
+
+
+   // The UTC unix timestamp seconds that a checkpoint expires at. Clients may update this value.
+   // If `checkpoint_expire_time_s` is older than now(), the checkpoint is considered expired.
+   // If `checkpoint_expire_time_s` is 0, the checkpoint will never expire.
+   checkpoint_expire_time_s: uint;
+
+
+   // Optional metadata associated with the checkpoint. For example, the writer can use this to clean up checkpoints from older writers.
+   metadata: CheckpointMetadata;
+}
+```
+
+### Public API
+
+We’ll make the following changes to the public API to support creating and using checkpoints:
+
+```
+/// Defines the scope targeted by a given checkpoint. If set to All, then the checkpoint will include
+/// all writes that were issued at the time that create_checkpoint is called. If force_flush is true,
+/// then SlateDB will force the current wal, or memtable if wal_enabled is false, to flush its data.
+/// Otherwise, the database will wait for the current wal or memtable to be flushed due to
+/// flush_interval or reaching l0_sst_size_bytes, respectively. If set to Durable, then the
+/// checkpoint includes only writes that were durable at the time of the call. This will be faster, but
+/// may not include data from recent writes.
+enum CheckpointScope {
+    All{force_flush: bool},
+    Durable
+}
+
+/// Specify options to provide when creating a checkpoint.
+struct CheckpointOptions {
+    scope: CheckpointScope
+}
+
+impl Db {
+    /// Opens a Db from a checkpoint. If no database This will create a new db under path from
+    /// the checkpoint ID specified in source_checkpoint. SlateDB will look for the manifest
+    /// containing the checkpoint under source_path. The newly created database is a shallow
+    /// copy of the source database, and contains all the data from source_checkpoint. New writes
+    /// will be written to the new database and will not be reflected in the source database.
+    pub async fn open_from_checkpoint(
+        path: Path,
+        object_store: Arc<dyn ObjectStore>,
+        source_path: Path,
+        source_checkpoint: UUID,
+    ) -> Result<Self, SlateDBError> {
+        …
+    }
+
+    /// Create a checkpoint using the provided options. Returns the ID of the created checkpoint.
+    pub async fn create_checkpoint(options: &CheckpointOptions) -> Result<uuid::UUID, SlateDBError> {
```

Comment:
It would be good to return the manifest id as well, since it could take multiple manifest write attempt attempts before this succeeds.